File: source/_posts/Algorithm/1.leetcode/数据结构/二叉树/二叉搜索树/96.不同的二叉搜索树.py
#
# @lc app=leetcode.cn id=96 lang=python3
#
# [96] 不同的二叉搜索树
#
# f(n) = f(n-1)*f(0) + f(n-2)*f(1) +……f(0)*f(n-1)


# @lc code=start
class Solution:
    def numTrees2(self, n: int) -> int:
        if n <= 1:
            return 1
        res = 0
        # 并没有栈溢出，因此我们考虑使用 hashmap 来优化
        for i in range(1, n + 1):
            res += self.numTrees2(i - 1) * self.numTrees2(n - i)
        return res

    visited = dict()

    def numTrees(self, n: int) -> int:
        if n in self.visited:
            return self.visited.get(n)
        if n <= 1:
            return 1
        res = 0
        for i in range(1, n + 1):
            res += self.numTrees(i - 1) * self.numTrees(n - i)
        self.visited[n] = res
        return res


# @lc code=end

File: source/_posts/Algorithm/数据结构/二叉树/二叉搜索树/96.不同的二叉搜索树.java
/*
 * @lc app=leetcode.cn id=96 lang=java
 *
 * [96] 不同的二叉搜索树
 */

// @lc code=start
class SolutionBST {
    /**
     * 递归
     * 
     * @param n
     * @return
     */
    public int numTrees(int n) {
        if (n <= 1) {
            return 1;
        }
        int res = 0;
        for (int i = 1; i <= n; i++) {
            res = numTrees(i) * numTrees(n - i - 1);
        }
        return res;
    }

    /**
     * 迭代
     * 
     * @param n
     * @return
     */
    public int countBst(int n) {
        int[] res = new int[n + 1];
        res[0] = 1;
        for (int index = 1; index < res.length; index++) {
            int tmp = 0;
            for (int j = 0; j < index; j++) {
                tmp += res[j] * res[index - j - 1];
            }
            res[index] = tmp;
        }
        return res[n];
    }

    public static void main(String[] args) {
        SolutionBST test = new SolutionBST();
        test.countBst(2);
    }

}
// @lc code=end
